Query parameter validation using Literal types in a dataclass fails

Query parameter validation fails to coerce valid data when using `Literal` in a dataclass:

```python
import json
from dataclasses import dataclass
from typing import Literal

from sanic import Sanic
from sanic.response import empty
from sanic_ext import validate


app = Sanic(name='name')


@dataclass
class Spec:
    name: Literal['foo', 'bar']


@app.get('/route')
@validate(query=Spec)
def route(_, body: Spec):
    print(body.name)
    return empty()


req, resp = app.test_client.get('/route', params={'name': 'baz'})
# ✔️ Validation catches invalid data:
# sanic_ext.exceptions.ValidationError: Invalid request body: Spec. Error: Value '['baz']' must be one of foo, bar

req, resp = app.test_client.get('/route', params={'name': 'bar'})
# ❌ Validation error coercing valid data:
# [ERROR] Exception occurred while handling uri: 'http://127.0.0.1:39541/route?name=bar'
# Traceback (most recent call last):
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/validators.py", line 25, in validate_body
#     return validator(model, body)
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/validators.py", line 38, in _validate_annotations
#     return check_data(model, body, schema, allow_multiple, allow_coerce)
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/check.py", line 147, in check_data
#     hydration_values[key] = hint.validate(
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/check.py", line 100, in validate
#     value = self.coerce(value)
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/check.py", line 109, in coerce
#     coerce_types = [arg for arg in args if not isinstance(None, arg)]
#   File ".../env/lib/python3.8/site-packages/sanic_ext/extras/validation/check.py", line 109, in <listcomp>
#     coerce_types = [arg for arg in args if not isinstance(None, arg)]
# TypeError: isinstance() arg 2 must be a type or tuple of types
# ...
# sanic_ext.exceptions.ValidationError: Invalid request body: Spec. Error: isinstance() arg 2 must be a type or tuple of types


```

**Expected behavior**
No error when valid query parameters are used.

**Environment (please complete the following information):**
Sanic v22.12.0, sanic-routing==22.8.0, sanic-testing==22.9.0, sanic-ext==22.12.0

--- a/sanic_ext/extras/validation/check.py
+++ b/sanic_ext/extras/validation/check.py
@@ -112,9 +112,11 @@ class Hint(NamedTuple):
 
     def coerce(self, value):
         if is_generic(self.coerce_type):
-            if get_origin(self.coerce_type) == Literal:
-                return value
             args = get_args(self.coerce_type)
+            if get_origin(self.coerce_type) == Literal or (
+                all(get_origin(arg) == Literal for arg in args)
+            ):
+                return value
             if type(None) in args and value is None:
                 return None
             coerce_types = [arg for arg in args if not isinstance(None, arg)]

--- a/sanic_ext/extras/validation/check.py
+++ b/sanic_ext/extras/validation/check.py
@@ -1,7 +1,16 @@
 from __future__ import annotations
 
 from dataclasses import _HAS_DEFAULT_FACTORY  # type: ignore
-from typing import Any, Literal, NamedTuple, Optional, Tuple, Union, get_args
+from typing import (
+    Any,
+    Literal,
+    NamedTuple,
+    Optional,
+    Tuple,
+    Union,
+    get_args,
+    get_origin,
+)
 
 from sanic_ext.utils.typing import UnionType, is_generic, is_optional
 
@@ -103,6 +112,8 @@ class Hint(NamedTuple):
 
     def coerce(self, value):
         if is_generic(self.coerce_type):
+            if get_origin(self.coerce_type) == Literal:
+                return value
             args = get_args(self.coerce_type)
             if type(None) in args and value is None:
                 return None

--- a/tests/extra/test_validation.py
+++ b/tests/extra/test_validation.py
@@ -1,0 +1,47 @@
+from dataclasses import dataclass
+from sys import version_info
+from typing import Literal, Union
+
+import pytest
+from sanic import text
+
+from sanic_ext import validate
+
+
+@pytest.mark.parametrize(
+    "annotation",
+    (
+        (
+            Literal["foo"],
+            Literal["foo", "bar"],
+            Union[Literal["foo"], Literal["bar"]],
+        )
+    ),
+)
+def test_literal(app, annotation):
+    @dataclass
+    class Spec:
+        name: annotation
+
+    @app.get("/")
+    @validate(query=Spec)
+    def route(_, query: Spec):
+        return text(query.name)
+
+    _, response = app.test_client.get("", params={"name": "foo"})
+    assert response.text == "foo"
+
+
+@pytest.mark.skipif(version_info < (3, 9), reason="Not needed on 3.8")
+def test_literal_3_9(app):
+    @dataclass
+    class Spec:
+        name: Literal["foo"] | Literal["bar"]
+
+    @app.get("/")
+    @validate(query=Spec)
+    def route(_, query: Spec):
+        return text(query.name)
+
+    _, response = app.test_client.get("", params={"name": "foo"})
+    assert response.text == "foo"

--- a/tests/extra/test_validation.py
+++ b/tests/extra/test_validation.py
@@ -32,8 +32,8 @@ def test_literal(app, annotation):
     assert response.text == "foo"
 
 
-@pytest.mark.skipif(version_info < (3, 9), reason="Not needed on 3.8")
-def test_literal_3_9(app):
+@pytest.mark.skipif(version_info < (3, 10), reason="Not needed on 3.10")
+def test_literal_3_10(app):
     @dataclass
     class Spec:
         name: Literal["foo"] | Literal["bar"]

